### ISSUE TITLE
Resource display UI in construction site window

### ISSUE THREAD

**Issue by bunnybot** [widelands/widelands#1245]
The construction site windows currently show the remaining required resources, as icons, with the preference buttons to the right. Once all wares of one type are delivered, the row is basically empty besides the preferences buttons. This looks confusing.

I suggest to initially show the demand for wares in greyed-out-fashion and replace them by the regular icon when a ware is delivered, similar to the display in finished buildings. This way, the layout of the window would not change.

Additionally, one could put the preference buttons to the left so that they align.

------------------------------------
Imported from Launchpad using lp2gh.

 * date created: 2010-10-31T18:36:20Z
 * owner: nomeata
 * the launchpad url was https://bugs.launchpad.net/bugs/669192

**Comment by bunnybot** [comment]
(by sirver)
I find this very intuitive. But I put it up for a vote.  I agree that the preferences buttons could be left.

+1 for only moving the buttons to the left

**Comment by bunnybot** [comment]
(by nomeata)
Just for clarification: What do you mean by “this”? My proposal, or the current state?

**Comment by bunnybot** [comment]
(by nomeata)
Ah, sorry, your last sentence makes it clearer...

In that case: How about removing the empty rows of fulfilled resources? I especially don’t like the rows that only contain buttons but nothing else?

**Comment by bunnybot** [comment]
(by hjd)
What I like about the current dialog is that it shows:
-how many you need of a resource (greyed out)
-how many you have and haven't been spent in the building process yet (normal color)
-when an item is spent it disappears decrementing the row by one
The last one makes it easier to see the remaining cost of the building. The proposed scheme would only show resource you need, and resources you have. Though in this case I wouldn't know if they are already used in the building or "piled up outside" ready to be used. This would, IMO, provide me less information than the current dialog.
On the other hand, the current dialog makes it is harder to assess the total cost of the building if you check the building dialog before deciding whether or not you have resources to build another.

As for moving the buttons to the left: maybe. It will make them slightly inconsistent with how the other priority-buttons are placed, though one could argue there is enough difference between construction sites and buildings. I haven't made up my mind on this one yet.

Don't get me wrong, I find it a bit annoying that construction sites are resizing the window as it is updated and the empty rows don't seem to do much. One idea (sort of a counter-proposal) would be to remove rows with wares that have been used in the building and are now empty. This way, when an item is spent it is removed from the row, and when all items of that type is spent, the entire row is removed. Again, as mentioned above, this makes it slightly harder to calculate total cost.

**Comment by bunnybot** [comment]
(by nomeata)
Hmm, maybe I need to look at the window some more. I wasn’t ware that there are three states (missing resource, delivered  resource, used resource) shown differently (greyed out, normal color, invisible), and assumed that delivered resourced would be used immediately.

In that case, I guess I’d just prefer having empty rows removed and the preference buttons aligned, either on the left or on the far right, but in one column.

One could argue that the three states should all be visible, if someone has a good idea how used items could be displayed. This had the advantage of a non-resizing window, which I find appealing.

**Comment by bunnybot** [comment]
(by sirver)
The empty rows have one more feature: when you hover over the "empty" line, a tooltip will tell you what ware has been listed there; this is sometimes useful information.

So, won't fix?

**Comment by bunnybot** [comment]
(by nomeata)
No, I’m not convinced. Especially for new users, building their first woodcutter or similar and then looking that the construction site after the stuff has been delivered, the window just looks broken: A thin window, ~6 very small buttons vertically aligned, with no apparent purpose. It might make sense if you saw a few such windows before, but it is neither user-friendly nor good-looking.

Having thought about this some more, I think the proper solution is really a three-valued visual display for each item: Missing, delivered, used. This will be more consistent, easier to understand and carrying more information.

The question then would be how to indicate wares that are used from wares that are delivered. Maybe a checkmark?

**Comment by bunnybot** [comment]
(by sirver)
I would be very sad to see the disappearing go. It is a very satisfying thing seeing the lines become shorter as your buildings grows more finished. How about just finishing the growing/shrinking of the window?

 Otherwise, I'd like to hear more voices on this.

**Comment by bunnybot** [comment]
(by sirver)
Btw: the buildings statistics menu is really confusing. And the ware production statistics window should have the same layout and visual as the stock window... just in case you are looking for stuff to work on :P.

**Comment by bunnybot** [comment]
(by nomeata)
Wouldn’t green checkmarks which appear over the used items be equally satisfying? Maybe I should create a mock-up.

**Comment by bunnybot** [comment]
(by ha-kripo)
Construction site window:

I also like the disappearing of wares. If there is no disappearing one ware after another, all wares would disappear at once (from the stock) when the construction site is finished. This would be a step backwards in my opinion. 
I'm not for a new visualization of a used ware, because it simply disappears (from the construction window and stock) like other wares consumed in production buildings. 

But I agree that empty rows with priority buttons make no sense. They should be removed.

I don't have a preference where the priority buttons are, but if they are at the left or at the far right in construction site windows, they should have the same place in production buildings.

**Comment by bunnybot** [comment]
(by nomeata)
Hmm, I don’t agree (and by now, I could support my opinion with some code), but I guess I stand alone on this...

(Note that the ware would still disappear from stock when used in the construction. It is just a change of display in the construction site window, where you’d see which resources had been used (indicated by a checkmark, or maybe crossed out, or something else), to make the window less surprising and more informative.)

I guess this can be marked Won’t fix.

**Comment by bunnybot** [comment]
(by sirver)
I just took a look at your code and I still prefer the current way of displaying wares queues. Your way does not provide the quick feedback of how much work is there still to do before the building is done. Seeing the remaining build wares to be consumed seems more intuitive to me and I feel it also carries more information.

**Comment by bunnybot** [comment]
(by nomeata)
The quickest feedback of total remaining work is, IMHO, the progress bar at the top. I don’t see in what way the current state carries more information. Especially, as the current state does not, besides the progress bar, give any kind of relative feedback: „Of this large number of resources, so-and-so many have already been worked with, so many are present and so many are missing.“

I still wonder if there is an alternative to the green checkmark that makes the dialog more useful and pretty to everyone. Some other idea I had were progress bars behind each row, marking how far construction using this resource has advanced. The bar would never progress further than the non-greyed-out icons, the larger the bars the farther the construction and once all bars are full the building is finished.

Anyways, you’ll find me debating until this is marked won’t fix :-)

**Comment by bunnybot** [comment]
(by qcumber-some)
We should probably see this wish in conjunction with the wish at https://bugs.launchpad.net/widelands/+bug/584134 and the to find a completely new way to display all information being available:

* Empty spot, request can not be fulfilled
* Empty spot, can be fulfilled by an item on the way (request fulfilled)
* Filled spot (ware available for consumption)
* Consumed item (construction site only; consumed items in production buildings are only interesting for statistics)

On one hand, one could ask, what makes a construction site special to any other building that it should display the consumed items. If we don't find that a construction site is special, we need to involve the missing information in both cases or in no case to achieve common visual appearance.

I agree that there s no primary use from the information how many wares have been consumed (even not for construction sites), however it may be good to know how expensive such a building has been (agree to #4). That could be displayed in *any* building regardless of the type. However this is not primary information and could very well be hidden behind a "Details" or Help button.

It also needs to be said that I'm always a little annoyed by construction site windows resizing themselves in one direction, and I would definitely speak against making them additionally resize in the second direction by removing resource rows which have been completely used up. The empty rows containing the priority buttons only are not really that irritating for newbies, as they probably will not open up the construction site window right after the last resource has been consumed, and for a building which progresses so slowly that you don't really see the percentage at the top of the window rising.

I had a look at the proposed window, and have to state that the green marks *do* almost confuse me. And they provide meaningless information, I don't care in which order the wares are consumed.

To make it consistent, using the marker scheme would then mean for the production buildings, that with each and every loop through the production site's program, every 'consume' would cause a green marker instead of removing the item from the in queue. On the end of the production site's program (after a sleep), the marked wares should then be removed from the queue, causing requests for new wares.
That would display more (and useful) information: How many resources are used in one cycle (it would for example show the strange wares usage for producing the Barbarian Rations in the Taverns).
Using this schema then means that construction sites marking the used resources do explain themselves.

I hope this long text does not confuse the readers, and I am sorry that I can't provide an easy solution but only a secondary look at the problem, with maybe a slightly broader view.

**Comment by bunnybot** [comment]
(by nha)
I don't think it makes sense to display the total amount of required wares in a construction site window. It would be far more logical and useful to provide that kind of information in FieldActionWindow. Otherwise, players have to build and destroy construction sites to get the information.

The (totally unrelated) suggestion of displaying whether or not a ware item is currently being en route (in Transfer) to the site sounds kind of cool, though.

**Comment by bunnybot** [comment]
(by sirver)
This goes totally wild here. I try to sum up:

1) Joachim suggest that wares that are consumed in a constructionsite do not vanish, but instead get marked optically in any way that they are consumed.

2) Joachim also suggest that the priority button are moved to the left so that they align.

3) Jens text got me confused :). He suggest planing the display of constructionsite along with the display for productionsite to make them similar. He also brings up the point that wares that are on route could get an extra symbol in the constructionsite. 

I disagree to 1) after having checked Joachims source, I also disagree with his statement in #14; the current dialog shows me rather quickly which wares are troublemakers and which are produced in an ample amount. I also think the current dialog is very intuitive; obviously it could use a help window.

2) I can live with that but this never bothered me before.

3) This suggestion is not part of this bug report. I'd rather discuss this together with a blueprint. I am not convinced that the added complexity is worth the information gained: in my games, I know exactly which wares are trouble makers, I do not need to look at productionsites for this.

Joachim, your original post did not get much support in the whole discussion, I therefore mark this bug as won't fix.

**Comment by bunnybot** [comment]
(by kyromaster)
I just had the exactly same idea as the usual reporter. What about discussing everything beyond this in seperate reports and marking this as wishlist?
To sum up, the preference buttons are moved to the left. In the construction site window they disappear once all wares are delivered, the overall size of the window always stays the same.
This would be a quick way to improve usability IMHO.
